neo/Oracle/Protocols/HTTP/OracleHTTPRequest.cs

```diff
@@ -0,0 +1,69 @@
+using Neo.IO;
+using System.IO;
+using System.Text;
+
+namespace Neo.Oracle.Protocols.HTTP
+{
+    public class OracleHTTPRequest : OracleRequest
+    {
+        public enum HTTPMethod : byte
+        {
+            GET = 0,
+            POST = 1,
+            PUT = 2,
+            DELETE = 3
+        }
```

Comment:
I think we can support `GET` only.

Reply:
For us is easy to allow multiple methods, why we should limit the others?

Reply:
For us is easy to allow multiple methods, why we should limit this?

Reply:
The built-in Oracle should be as simple as possible, don't do a lot of complicated things. `PUT` and `DELETE` are rarely used. The problem with `POST` is that we may also need to submit additional data, we also need to encode the data, and then the encoding must be consistent. This will add too much complexity. In fact, `GET` can already meet most of the needs.

Reply:
`GET` is good for me, but I think that NeoFS require more than this

Reply:
POST is used as base method for JSON RPC and most REST APIs. Oracles would be used for integration with external systems, sot supporting full set of HTTP methods would be beneficial. At least GET and POST, as most widely used, should be there for sure, aren't they?

Reply:
> `GET` is good for me, but I think that NeoFS require more than this

NeoFS needs to be implemented as one more protocol Oracle, implementing [own set of gRPC requests](https://github.com/nspcc-dev/neofs-proto)

Reply:
HTTP POST semantically, its function is to submit data to the server. So theoretically the same data should only be submitted once. But our Oracle is a distributed system with multiple nodes doing the same job. In this case, I don't think POST is still following the original semantics. Oracle should only get the data, in which case GET is sufficient.

As for NeoFS, it must be decoupled from Oracle. I don't think we should allow calling NeoFS gRPC in smart contracts. We should provide separate SYSCALLs for NeoFS.

Reply:
Removed `PUT` and `DELETE` , time to discuss `POST`

Reply:
For protocols on top of HTTP, POST verb semantically is the way to pass the request. For example in JSON RPC specification and in most RESTful APIs. So it's required for getting data from most of modern web services. We need at least both POST and GET to communicate with external world to get data.

Reply:
> As for NeoFS, it must be decoupled from Oracle.

Oracle is the interaction protocol between blockchain and external entities. NeoFS is an external entity. This is the reason why it is efficient to store data in distributed storage and not in blockchain itself. Consider getting data from the internet web page and from the NeoFS - both cases require interaction outside of the blockchain. Oracle designed to solve exactly this issue, so maybe there is no need to decouple NeoFS from the Oracle.

Various protocol support (http-oracle and neofs-oracle for example) will also confirm viability of Oracle protocol.

Reply:
> HTTP POST semantically, its function is to submit data to the server. So theoretically the same data should only be submitted once. But our Oracle is a distributed system with multiple nodes doing the same job. In this case, I don't think POST is still following the original semantics. Oracle should only get the data, in which case GET is sufficient.

Not exactly. Even if we're to look at the [old RFC 2616](https://tools.ietf.org/html/rfc2616#section-9.5) that says this:
> The POST method is used to request that the origin server accept the entity enclosed in the request as a new subordinate of the resource identified by the Request-URI in the Request-Line.

it at the same time says that

> The actual function performed by the POST method is determined by the server and is usually dependent on the Request-URI.

And more modern versions of the standard ([RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.2.2)) are even more liberal in their interpretation of `POST` (or just following the line of what is really done via `POST` on the net):

>  The POST method requests that the target resource process the representation enclosed in the request according to the resource's own specific semantics.

So it basically moves these questions to the upper layer and we have quite a lot of upper layers on top of HTTP these days.

Like JSON RPC, let's say we're making a bet with @realloc on whether there will be a 0.70.0 release of neo-go before the end of November. For a release proof we need to make a JSON RPC `getversion` call to our mainnet nodes, so we need to be able to do that via our new shiny Oracles and if they can't do a `POST` we're out of luck.

Also note that while in our bet example we could probably arrange some `GET`-able proof for the matter, many of the endpoints providing data may be outside of the smart contract developer control and may not have any other means to get data but making a `POST` request.

So I think there is no point in limiting our HTTP Oracles to just `GET` requests, support for other methods doesn't break any standard compliance and there are useful real-world scenarios that require us to have broader spectrum of supported methods.

Reply:
Simple. Decoupled. These two points are my most basic requirements for both Oracle and NeoFS.

Simple, the simpler the better. No `POST`, only `GET`.

Decoupled. I need NeoFS to work without Oracle.

Reply:
If NeoFS will work not through Oracle protocol but a simple syscall it may halt the execution of SC while waiting for NeoFS response. Maybe it would be more reliable to have all interactions with external systems go via a single Oracle protocol. It could be the simplest solution in the end.

Reply:
@realloc Can you create a new issue and list the features that NeoFS will offer to smart contracts? (I think it may be a SYSCALL list.) In this way, we can easily discuss the specific implementation and whether it can be decoupled.